src/main/java/net/minecraftforge/network/HandshakeMessages.java

```diff
@@ -68,7 +69,7 @@ public int getAsInt() {
 
         public S2CModList()
         {
-            this.mods = ModList.get().getMods().stream().map(IModInfo::getModId).collect(Collectors.toList());
+            this.mods = ModList.get().getMods().stream().map(mod -> mod.getModId() + "#" + mod.getVersion().toString()).collect(Collectors.toList());
```

Comment:
This is not backwards compatible. Old clients do not understand this format.

Reply:
Sending the version within the mod id string was the only reasonable way I could think of for sending the mod version without changing the packet (if you can think of a better one, I'm open for suggestions). Old clients won't crash when receiving this packet, they will just store the `modid#version` string, which needs to be handled specially by mods who use `getModList()`. That, in my opinion, is not too bad of a trade-off for sending the mod version without a crash.

Reply:
You could send another packet as prefix to this one.

Reply:
in case you mean an entirely new packet for sending the versions to the client: I don't know if I'm a fan of adding yet another packet for this feature to work, especially if it's not needed (since the current approach works without it in a semi-backwards-compatible way, as described above). Of course, if you think that old clients not understanding the new server's format is a problem to that extent that the addition of another packet (or at that point, an edit in `S2CModList` because it'd be as backwards incompatible as a new packet) is warranted, I'll try to do that - though then this PR is probably not gonna get merged until 1.19.

Reply:
Sure, backwards compatibility is important.
But clients and servers should always share the same forge version. I know, there are some who don't do that.
But updating to a newer version of forge isn't magic, just updating it in the launcher of choice or downloading a newer version from the forge site.
Would be sad to wait for 1.19 for this wonderful new feature.

Reply:
Backwards compatibility is an important part of Forge and I doubt this PR will be merged in 1.18 without being backwards compatible.

A new packet is not backwards incompatible, it will simply be ignored by older clients. Your current change is not even "semi backwards compatible", because it changes the data in NetworkHooks.getConnectionData

Reply:
Wait, so a packet sent from the server that doesn't get recorgnized by the client (for example, if the client is outdated) does not cause the client to crash/not be able to complete the handshake? I thought that adding a new packet would inevitably make this PR a breaking change for that reason. If that's not the case, sorry for the misunderstanding, I will definitely use a packet for sending the data from the server to the client

Reply:
See `ClientHandshakePacketListenerImpl#handleCustomQuery` and `NetworkHooks.onCustomPayload`. If I read the code correctly, an unrecognized packet should not do anything harmful.